[Liquid Glass] Websites that have fixed/sticky headers after scrolling should not extend color into pocket

whsieh opened the pull request:
#### c9f1bebd8e25c4d0c16472d4c046be474632f23e
<pre>
[Liquid Glass] Websites that have fixed/sticky headers after scrolling should not extend color into pocket
<a href="https://bugs.webkit.org/show_bug.cgi?id=295569">https://bugs.webkit.org/show_bug.cgi?id=295569</a>
<a href="https://rdar.apple.com/155267492">rdar://155267492</a>

Reviewed by Abrar Rahman Protyasha.

Add support for a new internal setting, `TopContentInsetBackgroundCanChangeAfterScrolling`, that
controls whether or not the top sampled fixed/sticky element background color is allowed to change
as the user scrolls the page. By default, this is only enabled on iPhone (based on a user interface
idiom check).

* LayoutTests/fast/page-color-sampling/color-sampling-for-sticky-element-after-scrolling.html:

Force the flag on in this layout test, whose purpose is to check that we&apos;re able to resample when
scrolling. This test still has user-facing impact when running under iPhone UI idiom.

* LayoutTests/fast/page-color-sampling/no-resampling-after-scrolling-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/no-resampling-after-scrolling.html: Added.

Add a new layout test that exercises this behavior, by forcing the flag off and scrolling with a
sticky-positioned header.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async waitForZoomingOrScrollingToEnd):
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::reset):
(WebCore::LocalFrameView::show):
(WebCore::LocalFrameView::setLastUserScrollType):

Add a flag to keep track of whether we&apos;ve observed _any_ user-driven explicit scrolling since the
page load committed.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateFixedContainerEdges):

Implement the main fix here — in the case where:

1.  `TopContentInsetBackgroundCanChangeAfterScrolling` is disabled (i.e. macOS or iPadOS).
2.  The main frame has been scrolled explicitly by the user.
3.  Document load has finished (i.e. the main document is no longer parsing).

...avoid resampling the top viewport edge; this naturally falls back to the last sampled color, or
no color at all if we haven&apos;t seen a fixed element near the top edge yet.

* Source/WebKit/Shared/Cocoa/WebPreferencesDefaultValuesCocoa.mm:
(WebKit::defaultTopContentInsetBackgroundCanChangeAfterScrolling):
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultTopContentInsetBackgroundCanChangeAfterScrolling):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:

Canonical link: <a href="https://commits.webkit.org/297118@main">https://commits.webkit.org/297118@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a16495beff344a4a2071138458d14b83ed4ad008

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110596 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30255 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20688 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116622 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60863 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112559 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38844 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84102 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113544 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24704 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99587 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64543 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24067 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17727 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60417 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103087 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/94090 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17786 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119412 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109150 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37636 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27951 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93068 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38010 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95858 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92890 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23668 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37904 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15647 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33613 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37531 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43002 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133425 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37194 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36052 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40533 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38902 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5309,17 +5309,25 @@ void Page::updateFixedContainerEdges(BoxSideSet sides)
         auto maximumOffset = frameView->maximumScrollOffset();
 
-        if (scrollOffset.y() < minimumOffset.y())
+        bool canSampleTopEdge = settings().topContentInsetBackgroundCanChangeAfterScrolling()
+            || !frameView->wasEverScrolledExplicitlyByUser()
+            || document->parsing();
```
The top edge can _not_ be sampled if all of the following:
- `TopContentInsetBackgroundCanChangeAfterScrolling` is enabled
- Main frame hasn't been explicitly scrolled by the user
- Document load has finished

Would that not read like this instead?


```suggestion
        bool canSampleTopEdge = settings().topContentInsetBackgroundCanChangeAfterScrolling()
            || frameView->wasEverScrolledExplicitlyByUser()
            || document->parsing();
```

whsieh replied:
Ah, good catch. So the code is correct, but the commit message is wrong — the second point should read:

> 2.  The main frame has been scrolled explicitly by the user.

(...instead of _never been scrolled_). Thanks for the review!

aprotyas replied:
Makes more sense!

whsieh committed:
[Liquid Glass] Websites that have fixed/sticky headers after scrolling should not extend color into pocket
https://bugs.webkit.org/show_bug.cgi?id=295569
rdar://155267492

Reviewed by Abrar Rahman Protyasha.

Add support for a new internal setting, `TopContentInsetBackgroundCanChangeAfterScrolling`, that
controls whether or not the top sampled fixed/sticky element background color is allowed to change
as the user scrolls the page. By default, this is only enabled on iPhone (based on a user interface
idiom check).

* LayoutTests/fast/page-color-sampling/color-sampling-for-sticky-element-after-scrolling.html:

Force the flag on in this layout test, whose purpose is to check that we're able to resample when
scrolling. This test still has user-facing impact when running under iPhone UI idiom.

* LayoutTests/fast/page-color-sampling/no-resampling-after-scrolling-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/no-resampling-after-scrolling.html: Added.

Add a new layout test that exercises this behavior, by forcing the flag off and scrolling with a
sticky-positioned header.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async waitForZoomingOrScrollingToEnd):
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::reset):
(WebCore::LocalFrameView::show):
(WebCore::LocalFrameView::setLastUserScrollType):

Add a flag to keep track of whether we've observed _any_ user-driven explicit scrolling since the
page load committed.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateFixedContainerEdges):

Implement the main fix here — in the case where:

1.  `TopContentInsetBackgroundCanChangeAfterScrolling` is disabled (i.e. macOS or iPadOS).
2.  The main frame has been scrolled explicitly by the user.
3.  Document load has finished (i.e. the main document is no longer parsing).

...avoid resampling the top viewport edge; this naturally falls back to the last sampled color, or
no color at all if we haven't seen a fixed element near the top edge yet.

* Source/WebKit/Shared/Cocoa/WebPreferencesDefaultValuesCocoa.mm:
(WebKit::defaultTopContentInsetBackgroundCanChangeAfterScrolling):
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultTopContentInsetBackgroundCanChangeAfterScrolling):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:

Canonical link: https://commits.webkit.org/297118@main
